pkg/oauthserver/authenticator/password/keystonepassword/keystonepassword.go

```diff
@@ -18,22 +20,55 @@ import (
 
 // keystonePasswordAuthenticator uses OpenStack keystone to authenticate a user by password
 type keystonePasswordAuthenticator struct {
-	providerName   string
-	url            string
-	client         *http.Client
-	domainName     string
-	identityMapper authapi.UserIdentityMapper
+	providerName           string
+	url                    string
+	client                 *http.Client
+	domainName             string
+	identityMapper         authapi.UserIdentityMapper
+	useKeystoneIdentity    bool
 }
 
 // New creates a new password authenticator that uses OpenStack keystone to authenticate a user by password
 // A custom transport can be provided (typically to customize TLS options like trusted roots or present a client certificate).
 // If no transport is provided, http.DefaultTransport is used
-func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper) authenticator.Password {
+func New(providerName string, url string, transport http.RoundTripper, domainName string, identityMapper authapi.UserIdentityMapper, useKeystoneIdentity bool) authenticator.Password {
 	if transport == nil {
 		transport = http.DefaultTransport
 	}
 	client := &http.Client{Transport: transport}
-	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper}
+	return &keystonePasswordAuthenticator{providerName, url, client, domainName, identityMapper, useKeystoneIdentity}
+}
+
+// Authenticate user and return his Keystone ID
+func getUserIDv3(client *gophercloud.ProviderClient, options gophercloud.AuthOptions) (string, error) {
+	// Override the generated service endpoint with the one returned by the version endpoint.
+	v3Client := openstack.NewIdentityV3(client)
+
+	var body struct {
+		Token struct {
+			User struct {
+				ID  string
+			} `mapstructure:"user"`
+		} `mapstructure:"token"`
+	}
+
+	// Issue new unscoped token
+	var scope *tokens3.Scope
+	result := tokens3.Create(v3Client, options, scope)
+
+	_, err := result.ExtractToken()
```

Comment:
why call ExtractToken and not use the value?

Reply:
I just didn't want to duplicate code :) ExtractToken validates that there was not an error, I used it for this reason. I'll rewrite the code as you ask.